// @gravity-ui/chartkit/build/plugins/highcharts/renderer/helpers/constants.js
export var HighchartsType;
(function (HighchartsType) {
    HighchartsType["Area"] = "area";
    HighchartsType["Arearange"] = "arearange";
    HighchartsType["Bar"] = "bar";
    HighchartsType["Boxplot"] = "boxplot";
    HighchartsType["Bubble"] = "bubble";
    HighchartsType["Column"] = "column";
    HighchartsType["Columnrange"] = "columnrange";
    HighchartsType["Funnel"] = "funnel";
    HighchartsType["Heatmap"] = "heatmap";
    HighchartsType["Line"] = "line";
    HighchartsType["Map"] = "map";
    HighchartsType["Networkgraph"] = "networkgraph";
    HighchartsType["Pie"] = "pie";
    HighchartsType["Sankey"] = "sankey";
    HighchartsType["Scatter"] = "scatter";
    HighchartsType["Streamgraph"] = "streamgraph";
    HighchartsType["Timeline"] = "timeline";
    HighchartsType["Treemap"] = "treemap";
    HighchartsType["Variwide"] = "variwide";
    HighchartsType["Waterfall"] = "waterfall";
    HighchartsType["Wordcloud"] = "wordcloud";
    HighchartsType["Xrange"] = "xrange";
})(HighchartsType || (HighchartsType = {}));
export var NavigatorLinesMode;
(function (NavigatorLinesMode) {
    NavigatorLinesMode["All"] = "all";
    NavigatorLinesMode["Selected"] = "selected";
})(NavigatorLinesMode || (NavigatorLinesMode = {}));
export const DEFAULT_LINES_LIMIT = 50;
