// @gravity-ui/chartkit/build/plugins/d3/examples/bar-x/Basic.js
import React from 'react';
import { groups } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
function prepareData({ field, filterNulls } = {
    field: 'platform',
}) {
    const gamesByPlatform = groups(nintendoGames, (item) => item[field]);
    let resultData = gamesByPlatform;
    if (filterNulls) {
        resultData = gamesByPlatform.filter(([value]) => typeof value === 'number');
    }
    const data = resultData.map(([value, games]) => ({
        x: value,
        y: games.length,
    }));
    return {
        categories: resultData.map(([key]) => key),
        series: [
            {
                data,
                name: 'Games released',
            },
        ],
    };
}
export const BasicBarXChart = () => {
    const { categories, series } = prepareData();
    const widgetData = {
        series: {
            data: series.map((s) => ({
                type: 'bar-x',
                data: s.data,
                name: s.name,
            })),
        },
        xAxis: {
            type: 'category',
            categories: categories.map(String),
            title: {
                text: 'Game Platforms',
            },
        },
        yAxis: [{ title: { text: 'Number of games released' } }],
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
export const BasicLinearBarXChart = () => {
    const { series } = prepareData({ field: 'meta_score' });
    const widgetData = {
        series: {
            data: series.map((s) => ({
                type: 'bar-x',
                data: s.data,
                name: s.name,
            })),
        },
        xAxis: {
            title: {
                text: 'Meta scores',
            },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
export const BasicDateTimeBarXChart = () => {
    const { series } = prepareData({ field: 'date', filterNulls: true });
    const widgetData = {
        series: {
            data: series.map((s) => ({
                type: 'bar-x',
                data: s.data,
                name: s.name,
            })),
        },
        xAxis: {
            type: 'datetime',
            title: {
                text: 'Release date',
            },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
