// @gravity-ui/chartkit/build/plugins/d3/examples/bar-x/DataLabels.js
import React from 'react';
import { groups, sort } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
const years = Array.from(new Set(nintendoGames.map((g) => g.date ? new Date(g.date).getFullYear().toString() : 'unknown')));
function prepareData() {
    const games = sort(nintendoGames.filter((d) => {
        return d.date && d.user_score;
    }), (d) => d.date);
    const groupByYear = (d) => (d.date ? new Date(d.date).getFullYear() : 'unknown');
    const byGenre = (genre) => {
        const data = groups(games.filter((d) => d.genres.includes(genre)), groupByYear).map(([year, items]) => {
            return {
                x: years.indexOf(String(year)),
                y: items.length,
            };
        });
        return {
            type: 'bar-x',
            name: genre,
            dataLabels: {
                enabled: true,
            },
            stacking: 'normal',
            data,
        };
    };
    return [byGenre('Strategy'), byGenre('Shooter'), byGenre('Puzzle'), byGenre('Action')];
}
export const DataLabels = () => {
    const series = prepareData();
    const widgetData = {
        series: {
            data: series,
        },
        xAxis: {
            categories: years,
            type: 'category',
            title: {
                text: 'Release year',
            },
            ticks: { pixelInterval: 200 },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
