// @gravity-ui/chartkit/build/plugins/d3/examples/bar-x/StackedColumns.js
import React from 'react';
import { groups } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
function prepareData() {
    const grouped = groups(nintendoGames, (d) => d.platform, (d) => (d.date ? new Date(d.date).getFullYear() : 'unknown'));
    const categories = [];
    const series = grouped.map(([platform, years]) => {
        return {
            name: platform,
            data: years.map(([year, list]) => {
                categories.push(String(year));
                return {
                    x: String(year),
                    y: list.length,
                };
            }),
        };
    });
    return { categories, series };
}
export const StackedColumns = () => {
    const { series, categories } = prepareData();
    const data = series.map((s) => {
        return {
            type: 'bar-x',
            stacking: 'normal',
            name: s.name,
            data: s.data,
        };
    });
    const widgetData = {
        series: {
            data: data,
        },
        xAxis: {
            type: 'category',
            categories: categories.sort(),
            title: {
                text: 'Release year',
            },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
