// @gravity-ui/chartkit/build/plugins/d3/renderer/hooks/useShapes/scatter/prepare-data.js
import get from 'lodash/get';
import { getXValue, getYValue } from '../utils';
const getFilteredLinearScatterData = (data) => {
    return data.filter((d) => typeof d.x === 'number' && typeof d.y === 'number');
};
export const prepareScatterData = (args) => {
    const { series, xAxis, xScale, yAxis, yScale } = args;
    return series.reduce((acc, s) => {
        const yAxisIndex = get(s, 'yAxis', 0);
        const seriesYAxis = yAxis[yAxisIndex];
        const seriesYScale = yScale[yAxisIndex];
        const filteredData = xAxis.type === 'category' || seriesYAxis.type === 'category'
            ? s.data
            : getFilteredLinearScatterData(s.data);
        filteredData.forEach((d) => {
            acc.push({
                point: {
                    data: d,
                    series: s,
                    x: getXValue({ point: d, xAxis, xScale }),
                    y: getYValue({ point: d, yAxis: seriesYAxis, yScale: seriesYScale }),
                    opacity: get(d, 'opacity', null),
                },
                hovered: false,
                active: true,
            });
        });
        return acc;
    }, []);
};
