// @gravity-ui/chartkit/build/plugins/d3/examples/bar-y/Basic.js
import React from 'react';
import { groups } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
function prepareData(field = 'platform') {
    const gamesByPlatform = groups(nintendoGames, (item) => item[field]);
    const data = gamesByPlatform.map(([value, games]) => ({
        y: value,
        x: games.length,
    }));
    return {
        categories: gamesByPlatform.map(([key]) => key),
        series: [
            {
                data,
                name: 'Games released',
            },
        ],
    };
}
export const Basic = () => {
    const { categories, series } = prepareData();
    const widgetData = {
        series: {
            data: series.map((s) => ({
                type: 'bar-y',
                data: s.data,
                name: s.name,
            })),
        },
        xAxis: { title: { text: 'Number of games released' } },
        yAxis: [
            {
                type: 'category',
                categories: categories.map(String),
                title: {
                    text: 'Game Platforms',
                },
            },
        ],
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
