// @gravity-ui/chartkit/build/plugins/d3/examples/bar-y/NegativeValues.js
import React from 'react';
import { dateTime } from '@gravity-ui/date-utils';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import marsWeatherData from '../mars-weather';
export const NegativeValues = () => {
    const data = marsWeatherData.map((d) => ({
        y: dateTime({ input: d.terrestrial_date, format: 'YYYY-MM-DD' }).valueOf(),
        x: d.min_temp,
    }));
    const widgetData = {
        series: {
            data: [
                {
                    type: 'bar-y',
                    data: data,
                    name: 'Min temperature',
                },
            ],
        },
        xAxis: {
            title: {
                text: 'Min temperature',
            },
        },
        yAxis: [
            {
                type: 'datetime',
                title: {
                    text: 'Terrestrial date',
                },
            },
        ],
        title: {
            text: 'Mars weather',
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
