// @gravity-ui/chartkit/build/plugins/d3/examples/pie/Donut.js
import React from 'react';
import { groups } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
function prepareData() {
    const gamesByPlatform = groups(nintendoGames, (d) => d.esrb_rating || 'unknown');
    return gamesByPlatform.map(([value, games]) => ({
        name: value,
        value: games.length,
    }));
}
export const Donut = () => {
    const data = prepareData();
    const widgetData = {
        series: {
            data: [
                {
                    type: 'pie',
                    innerRadius: '50%',
                    data: data,
                },
            ],
        },
        legend: { enabled: true },
        title: {
            text: 'ESRB ratings',
            style: { fontSize: '12px', fontWeight: 'normal' },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
