// @gravity-ui/date-utils/build/settings/locales.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.localeLoaders = void 0;
exports.localeLoaders = {
    af: function () { return import('dayjs/locale/af.js'); },
    am: function () { return import('dayjs/locale/am.js'); },
    'ar-dz': function () { return import('dayjs/locale/ar-dz.js'); },
    'ar-iq': function () { return import('dayjs/locale/ar-iq.js'); },
    'ar-kw': function () { return import('dayjs/locale/ar-kw.js'); },
    'ar-ly': function () { return import('dayjs/locale/ar-ly.js'); },
    'ar-ma': function () { return import('dayjs/locale/ar-ma.js'); },
    'ar-sa': function () { return import('dayjs/locale/ar-sa.js'); },
    'ar-tn': function () { return import('dayjs/locale/ar-tn.js'); },
    ar: function () { return import('dayjs/locale/ar.js'); },
    az: function () { return import('dayjs/locale/az.js'); },
    be: function () { return import('dayjs/locale/be.js'); },
    bg: function () { return import('dayjs/locale/bg.js'); },
    bi: function () { return import('dayjs/locale/bi.js'); },
    bm: function () { return import('dayjs/locale/bm.js'); },
    'bn-bd': function () { return import('dayjs/locale/bn-bd.js'); },
    bn: function () { return import('dayjs/locale/bn.js'); },
    bo: function () { return import('dayjs/locale/bo.js'); },
    br: function () { return import('dayjs/locale/br.js'); },
    bs: function () { return import('dayjs/locale/bs.js'); },
    ca: function () { return import('dayjs/locale/ca.js'); },
    cs: function () { return import('dayjs/locale/cs.js'); },
    cv: function () { return import('dayjs/locale/cv.js'); },
    cy: function () { return import('dayjs/locale/cy.js'); },
    da: function () { return import('dayjs/locale/da.js'); },
    'de-at': function () { return import('dayjs/locale/de-at.js'); },
    'de-ch': function () { return import('dayjs/locale/de-ch.js'); },
    de: function () { return import('dayjs/locale/de.js'); },
    dv: function () { return import('dayjs/locale/dv.js'); },
    el: function () { return import('dayjs/locale/el.js'); },
    'en-au': function () { return import('dayjs/locale/en-au.js'); },
    'en-ca': function () { return import('dayjs/locale/en-ca.js'); },
    'en-gb': function () { return import('dayjs/locale/en-gb.js'); },
    'en-ie': function () { return import('dayjs/locale/en-ie.js'); },
    'en-il': function () { return import('dayjs/locale/en-il.js'); },
    'en-in': function () { return import('dayjs/locale/en-in.js'); },
    'en-nz': function () { return import('dayjs/locale/en-nz.js'); },
    'en-sg': function () { return import('dayjs/locale/en-sg.js'); },
    'en-tt': function () { return import('dayjs/locale/en-tt.js'); },
    en: function () { return import('dayjs/locale/en.js'); },
    eo: function () { return import('dayjs/locale/eo.js'); },
    'es-do': function () { return import('dayjs/locale/es-do.js'); },
    'es-mx': function () { return import('dayjs/locale/es-mx.js'); },
    'es-pr': function () { return import('dayjs/locale/es-pr.js'); },
    'es-us': function () { return import('dayjs/locale/es-us.js'); },
    es: function () { return import('dayjs/locale/es.js'); },
    et: function () { return import('dayjs/locale/et.js'); },
    eu: function () { return import('dayjs/locale/eu.js'); },
    fa: function () { return import('dayjs/locale/fa.js'); },
    fi: function () { return import('dayjs/locale/fi.js'); },
    fo: function () { return import('dayjs/locale/fo.js'); },
    'fr-ca': function () { return import('dayjs/locale/fr-ca.js'); },
    'fr-ch': function () { return import('dayjs/locale/fr-ch.js'); },
    fr: function () { return import('dayjs/locale/fr.js'); },
    fy: function () { return import('dayjs/locale/fy.js'); },
    ga: function () { return import('dayjs/locale/ga.js'); },
    gd: function () { return import('dayjs/locale/gd.js'); },
    gl: function () { return import('dayjs/locale/gl.js'); },
    'gom-latn': function () { return import('dayjs/locale/gom-latn.js'); },
    gu: function () { return import('dayjs/locale/gu.js'); },
    he: function () { return import('dayjs/locale/he.js'); },
    hi: function () { return import('dayjs/locale/hi.js'); },
    hr: function () { return import('dayjs/locale/hr.js'); },
    ht: function () { return import('dayjs/locale/ht.js'); },
    hu: function () { return import('dayjs/locale/hu.js'); },
    'hy-am': function () { return import('dayjs/locale/hy-am.js'); },
    id: function () { return import('dayjs/locale/id.js'); },
    is: function () { return import('dayjs/locale/is.js'); },
    'it-ch': function () { return import('dayjs/locale/it-ch.js'); },
    it: function () { return import('dayjs/locale/it.js'); },
    ja: function () { return import('dayjs/locale/ja.js'); },
    jv: function () { return import('dayjs/locale/jv.js'); },
    ka: function () { return import('dayjs/locale/ka.js'); },
    kk: function () { return import('dayjs/locale/kk.js'); },
    km: function () { return import('dayjs/locale/km.js'); },
    kn: function () { return import('dayjs/locale/kn.js'); },
    ko: function () { return import('dayjs/locale/ko.js'); },
    ku: function () { return import('dayjs/locale/ku.js'); },
    ky: function () { return import('dayjs/locale/ky.js'); },
    lb: function () { return import('dayjs/locale/lb.js'); },
    lo: function () { return import('dayjs/locale/lo.js'); },
    lt: function () { return import('dayjs/locale/lt.js'); },
    lv: function () { return import('dayjs/locale/lv.js'); },
    me: function () { return import('dayjs/locale/me.js'); },
    mi: function () { return import('dayjs/locale/mi.js'); },
    mk: function () { return import('dayjs/locale/mk.js'); },
    ml: function () { return import('dayjs/locale/ml.js'); },
    mn: function () { return import('dayjs/locale/mn.js'); },
    mr: function () { return import('dayjs/locale/mr.js'); },
    'ms-my': function () { return import('dayjs/locale/ms-my.js'); },
    ms: function () { return import('dayjs/locale/ms.js'); },
    mt: function () { return import('dayjs/locale/mt.js'); },
    my: function () { return import('dayjs/locale/my.js'); },
    nb: function () { return import('dayjs/locale/nb.js'); },
    ne: function () { return import('dayjs/locale/ne.js'); },
    'nl-be': function () { return import('dayjs/locale/nl-be.js'); },
    nl: function () { return import('dayjs/locale/nl.js'); },
    nn: function () { return import('dayjs/locale/nn.js'); },
    'oc-lnc': function () { return import('dayjs/locale/oc-lnc.js'); },
    'pa-in': function () { return import('dayjs/locale/pa-in.js'); },
    pl: function () { return import('dayjs/locale/pl.js'); },
    'pt-br': function () { return import('dayjs/locale/pt-br.js'); },
    pt: function () { return import('dayjs/locale/pt.js'); },
    rn: function () { return import('dayjs/locale/rn.js'); },
    ro: function () { return import('dayjs/locale/ro.js'); },
    ru: function () { return import('dayjs/locale/ru.js'); },
    rw: function () { return import('dayjs/locale/rw.js'); },
    sd: function () { return import('dayjs/locale/sd.js'); },
    se: function () { return import('dayjs/locale/se.js'); },
    si: function () { return import('dayjs/locale/si.js'); },
    sk: function () { return import('dayjs/locale/sk.js'); },
    sl: function () { return import('dayjs/locale/sl.js'); },
    sq: function () { return import('dayjs/locale/sq.js'); },
    'sr-cyrl': function () { return import('dayjs/locale/sr-cyrl.js'); },
    sr: function () { return import('dayjs/locale/sr.js'); },
    ss: function () { return import('dayjs/locale/ss.js'); },
    'sv-fi': function () { return import('dayjs/locale/sv-fi.js'); },
    sv: function () { return import('dayjs/locale/sv.js'); },
    sw: function () { return import('dayjs/locale/sw.js'); },
    ta: function () { return import('dayjs/locale/ta.js'); },
    te: function () { return import('dayjs/locale/te.js'); },
    tet: function () { return import('dayjs/locale/tet.js'); },
    tg: function () { return import('dayjs/locale/tg.js'); },
    th: function () { return import('dayjs/locale/th.js'); },
    tk: function () { return import('dayjs/locale/tk.js'); },
    'tl-ph': function () { return import('dayjs/locale/tl-ph.js'); },
    tlh: function () { return import('dayjs/locale/tlh.js'); },
    tr: function () { return import('dayjs/locale/tr.js'); },
    tzl: function () { return import('dayjs/locale/tzl.js'); },
    'tzm-latn': function () { return import('dayjs/locale/tzm-latn.js'); },
    tzm: function () { return import('dayjs/locale/tzm.js'); },
    'ug-cn': function () { return import('dayjs/locale/ug-cn.js'); },
    uk: function () { return import('dayjs/locale/uk.js'); },
    ur: function () { return import('dayjs/locale/ur.js'); },
    'uz-latn': function () { return import('dayjs/locale/uz-latn.js'); },
    uz: function () { return import('dayjs/locale/uz.js'); },
    vi: function () { return import('dayjs/locale/vi.js'); },
    'x-pseudo': function () { return import('dayjs/locale/x-pseudo.js'); },
    yo: function () { return import('dayjs/locale/yo.js'); },
    'zh-cn': function () { return import('dayjs/locale/zh-cn.js'); },
    'zh-hk': function () { return import('dayjs/locale/zh-hk.js'); },
    'zh-tw': function () { return import('dayjs/locale/zh-tw.js'); },
    zh: function () { return import('dayjs/locale/zh.js'); },
};
