// @gravity-ui/chartkit/build/plugins/d3/examples/line/LineWithMarkers.js
import React from 'react';
import { dateTime } from '@gravity-ui/date-utils';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
import nintendoGames from '../nintendoGames';
function prepareData() {
    const dataset = nintendoGames.filter((d) => d.date && d.user_score && new Date(d.date) > new Date(2022, 0, 1));
    const data = dataset.map((d) => ({
        x: d.date || undefined,
        y: d.user_score || undefined,
        custom: d,
    }));
    return {
        series: [
            {
                data,
                name: 'Nintendo games',
            },
        ],
    };
}
export const LineWithMarkers = () => {
    const { series } = prepareData();
    const widgetData = {
        series: {
            data: series.map((s) => ({
                type: 'line',
                data: s.data.filter((d) => d.x),
                name: s.name,
                marker: { enabled: true, symbol: 'square' },
            })),
        },
        yAxis: [
            {
                title: {
                    text: 'User score',
                },
            },
        ],
        xAxis: {
            type: 'datetime',
            title: {
                text: 'Release dates',
            },
        },
        tooltip: {
            renderer: (d) => {
                var _a;
                const point = (_a = d.hovered[0]) === null || _a === void 0 ? void 0 : _a.data;
                if (!point) {
                    return null;
                }
                const title = point.custom.title;
                const score = point.custom.user_score;
                const date = dateTime({ input: point.custom.date }).format('DD MMM YYYY');
                return (React.createElement(React.Fragment, null,
                    React.createElement("b", null, title),
                    React.createElement("br", null),
                    "Release date: ",
                    date,
                    React.createElement("br", null),
                    "User score: ",
                    score));
            },
        },
    };
    return (React.createElement(ExampleWrapper, null,
        React.createElement(ChartKit, { type: "d3", data: widgetData })));
};
