// @gravity-ui/chartkit/build/plugins/d3/examples/line/LogarithmicAxis.js
import React from 'react';
import { Col, Container, Row, Text } from '@gravity-ui/uikit';
import { randomNormal } from 'd3';
import { ChartKit } from '../../../../components/ChartKit';
import { ExampleWrapper } from '../ExampleWrapper';
export const LineWithLogarithmicAxis = () => {
    const randomY = randomNormal(0, 100);
    const widgetData = {
        series: {
            data: [
                {
                    type: 'line',
                    name: 'Line series',
                    data: new Array(25).fill(null).map((_, index) => ({
                        x: index,
                        y: Math.abs(randomY()),
                    })),
                },
            ],
        },
    };
    const lineWidgetData = Object.assign(Object.assign({}, widgetData), { title: { text: 'line' } });
    const logarithmicWidgetData = Object.assign(Object.assign({}, widgetData), { title: { text: 'logarithmic' }, yAxis: [
            {
                type: 'logarithmic',
            },
        ] });
    return (React.createElement(Container, { spaceRow: 5 },
        React.createElement(Row, { space: 1 },
            React.createElement(Text, { variant: "header-2" }, "logarithmic VS line")),
        React.createElement(Row, { space: 3 },
            React.createElement(Col, { s: 12, m: 6 },
                React.createElement(ExampleWrapper, null,
                    React.createElement(ChartKit, { type: "d3", data: lineWidgetData }))),
            React.createElement(Col, { s: 12, m: 6 },
                React.createElement(ExampleWrapper, null,
                    React.createElement(ChartKit, { type: "d3", data: logarithmicWidgetData }))))));
};
