// @gravity-ui/chartkit/build/constants/widget-data.js
export const SeriesType = {
    Area: 'area',
    BarX: 'bar-x',
    BarY: 'bar-y',
    Line: 'line',
    Pie: 'pie',
    Scatter: 'scatter',
    Treemap: 'treemap',
    Waterfall: 'waterfall',
};
export var DashStyle;
(function (DashStyle) {
    DashStyle["Dash"] = "Dash";
    DashStyle["DashDot"] = "DashDot";
    DashStyle["Dot"] = "Dot";
    DashStyle["LongDash"] = "LongDash";
    DashStyle["LongDashDot"] = "LongDashDot";
    DashStyle["LongDashDotDot"] = "LongDashDotDot";
    DashStyle["ShortDash"] = "ShortDash";
    DashStyle["ShortDashDot"] = "ShortDashDot";
    DashStyle["ShortDashDotDot"] = "ShortDashDotDot";
    DashStyle["ShortDot"] = "ShortDot";
    DashStyle["Solid"] = "Solid";
})(DashStyle || (DashStyle = {}));
export var SymbolType;
(function (SymbolType) {
    SymbolType["Circle"] = "circle";
    SymbolType["Diamond"] = "diamond";
    SymbolType["Square"] = "square";
    SymbolType["Triangle"] = "triangle";
    SymbolType["TriangleDown"] = "triangle-down";
})(SymbolType || (SymbolType = {}));
export var LineCap;
(function (LineCap) {
    LineCap["Butt"] = "butt";
    LineCap["Round"] = "round";
    LineCap["Square"] = "square";
    LineCap["None"] = "none";
})(LineCap || (LineCap = {}));
export var LayoutAlgorithm;
(function (LayoutAlgorithm) {
    LayoutAlgorithm["Binary"] = "binary";
    LayoutAlgorithm["Dice"] = "dice";
    LayoutAlgorithm["Slice"] = "slice";
    LayoutAlgorithm["SliceDice"] = "slice-dice";
    LayoutAlgorithm["Squarify"] = "squarify";
})(LayoutAlgorithm || (LayoutAlgorithm = {}));
